pennylane_lightning/core/src/simulators/lightning_qubit/gates/cpu_kernels/GateImplementationsLM.hpp

```diff
@@ -2062,33 +1954,33 @@ class GateImplementationsLM : public PauliGenerator<GateImplementationsLM> {
         all_wires.insert(all_wires.begin(), wires.begin(), wires.end());
         all_wires.insert(all_wires.begin(), controlled_wires.begin(),
                          controlled_wires.end());
-        std::vector<std::size_t> rev_wires(nw_tot);
-        std::vector<std::size_t> rev_wire_shifts(nw_tot);
-        for (std::size_t k = 0; k < nw_tot; k++) {
-            rev_wires[k] = (num_qubits - 1) - all_wires[(nw_tot - 1) - k];
-            rev_wire_shifts[k] = (one << rev_wires[k]);
-        }
+        const auto revs = reverseWires(num_qubits, all_wires, {});
+        const auto &rev_wires = revs.first;
         const std::vector<std::size_t> parity =
             Pennylane::Util::revWireParity(rev_wires);
-        const std::size_t dim = one << nw_tot;
+        const std::vector<std::size_t> indices =
+            Gates::generateBitPatterns(all_wires, num_qubits);
 
         std::size_t mask{0U};
         for (std::size_t k = 0; k < controlled_values.size(); k++) {
             mask |= static_cast<std::size_t>(controlled_values[n_contr - 1 - k])
                     << k;
         }
-        const std::size_t i0 = mask << one;
-        const std::size_t i1 = i0 | one;
+        const std::size_t i0 = indices[mask << one];
+        const std::size_t i1 = indices[(mask << one) | one];
         PL_LOOP_PARALLEL(1)
         for (std::size_t k = 0; k < exp2(num_qubits - nw_tot); k++) {
-            const auto &&indices = parity2indices(k, parity, rev_wire_shifts);
-            for (std::size_t i = 0; i < dim; i++) {
+            std::size_t offset{0U};
+            for (std::size_t i = 0; i < parity.size(); i++) {
+                offset |= ((k << i) & parity[i]);
+            }
+            for (std::size_t i = 0; i < indices.size(); i++) {
                 if ((i >> one) == mask) {
                     continue;
                 }
```

Comment:
The `if` condition usually will go positive or negative? :thinking:

Reply:
Not sure there is a way to tell.